cartographer/mapping_2d/sparse_pose_graph.cc

```diff
@@ -522,6 +531,34 @@ std::vector<SparsePoseGraph::Constraint> SparsePoseGraph::constraints() {
   return result;
 }
 
+void SparsePoseGraph::SetInitialTrajectoryPose(int from_trajectory_id,
+                                               int to_trajectory_id,
+                                               const transform::Rigid3d& pose,
+                                               const common::Time& time) {
+  common::MutexLocker locker(&mutex_);
+  initial_trajectory_poses_[from_trajectory_id] =
+      InitialTrajectoryPose{to_trajectory_id, pose, time};
+}
+
+transform::Rigid3d SparsePoseGraph::GetClosestTrajectoryPose(
+    int trajectory_id, const common::Time& time) {
+  common::MutexLocker locker(&mutex_);
+  CHECK(trajectory_nodes_.SizeOfTrajectoryOrZero(trajectory_id) > 0);
+  mapping::NodeId closest_node_id =
+      trajectory_nodes_.BeginOfTrajectory(trajectory_id)->id;
+  common::Duration min_time_diff = common::Duration::max();
+  for (const auto& node_id_data : trajectory_nodes_.trajectory(trajectory_id)) {
```

Comment:
How about:
1. Extract a pose interpolation function from `transform_interpolation_buffer.cc`
2. Use `lower_bound` to search efficiently for the nodes of interest. Maybe this can also be generalized somehow from `transform_interpolation_buffer.cc`?